### ISSUE TITLE
Improve handling of autograd.Function for training mode

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#93721]
The current handling of a user-defined `autograd.Function` is:
1) [Inline the forward() in eval() mode](https://github.com/facebookresearch/torchdynamo/blob/052b8dab6df5f027a34edb00dee73870433883d5/torchdynamo/variables/misc.py#L192)
2) [Break the graph in train() mode](https://github.com/facebookresearch/torchdynamo/blob/052b8dab6df5f027a34edb00dee73870433883d5/torchdynamo/variables/misc.py#L187)

This is not ideal.

A simple (but unsound) thing to do would be put the `autograd.Function` into the captured TorchDynamo FX graph.  This will be incorrect if the `autograd.Function` contains untraceable things, but works sometimes.

A better solution would be to have TorchDynamo and AOT Autograd cooperate to trace the `autograd.Function`.  We should:
1) Use TorchDynamo to trace the forwards
2) Use TorchDynamo to trace the backwards (need example inputs here)
3) If they both trace, construct a new autograd.Function to put in the generated FX graph.
4) Need some glue to connect this with AOT Autograd if there are side-effects,etc that dynamo needs to handle with Python code.

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @Chillee

**Comment by ngimel** [comment]
IIUC AutogradFunction is still a problem and causes graph breaks, triage review to discuss how we should treat it

**Comment by yunxing** [comment]
Is there any ETA for this issue? This is currently the biggest blocker that prevents us from using dynamo.

**Comment by jansel** [comment]
@yunxing one workaround would be to do:
```
@dynamo.allow_in_graph
def wrapper(x):
   return MyAutogradFunction.apply(x)
```

This will cause the `autograd.Function` to be traced (unsoundly).

@yanboliang might have an update on the status since it is assigned to him.

**Comment by Chillee** [comment]
@yunxing To be clear, as long as your autograd.Function is "traceable", using "allow_in_graph" should work. So this should be (more or less) as good as the existing `make_fx` handling of `autograd.Function`.

**Comment by yunxing** [comment]
@jansel @Chillee 
That makes sense, thanks. Yeah I'm aware of that option. One concern is people write crazy things in autograd.Function that may not be tracable.

**Comment by anijain2305** [comment]
@voznesenskym Can we close this one?
